stan/math/prim/arr/fun/matrix_gpu.hpp

```diff
@@ -25,169 +20,195 @@ enum copy_transposed_triangular {LOWER_TO_UPPER_TRIANGULAR = 0,
 namespace stan {
   namespace math {
 
+    enum triangularity {LOWER = 0, UPPER = 1, NONE = 2 };
+    enum copy_transposed_triangular {LOWER_TO_UPPER_TRIANGULAR = 0,
+      UPPER_TO_LOWER_TRIANGULAR = 1};
+
+    /**
+     * This class represents a matrix on the GPU. 
+     * 
+     * The matrix data is stored in the oclBuffer.     
+     * 
+     */
     class matrix_gpu {
       private:
-
         cl::Buffer oclBuffer;
 
       public:
+        int rows_;
+        int cols_;
+        int rows() {
+          return rows_;
+        }
 
-        int rows;
-        int cols;
-        cl::Buffer buffer() {
+        int cols() {
+          return cols_;
+        }
+
+        int size() {
+          return rows_ * cols_;
+        }
+
+        const cl::Buffer& buffer() const {
           return oclBuffer;
         }
 
-        //TODO: constructors with enumerator added when
-        // the matrix_gpu does not need to be READ_WRITE
-        matrix_gpu(int rows,  int cols) {
+        matrix_gpu() {
+          rows_ = 0;
+          cols_ = 0;
+        }
+        // TODO(Rok): constructors with enumerator added when
+        //  the matrix_gpu does not need to be READ_WRITE
+        /**
+         * Constructor for the matrix_gpu that 
+         * only allocates the buffer on the GPU.
+         * 
+         * @param rows number of matrix rows
+         * @param cols number of matrix columns
+         * 
+         * @throw <code>std::invalid_argument</code> if the
+         * matrices do not have matching dimensions
+         * 
+         */
+        matrix_gpu(int rows,  int cols)
```

Comment:
Would it be easier to wrap the OpenCL with try/catch directly, like `get_context()` and `cl::buffer()`?  Then it simplifies all the matrix code and scopes the try/catch appropriately.

Reply:
The get_context/get_queue just return the references if the context/queue was already initialized. Otherwise, init() is called, which has its own try/catch. I modified this to only wrap the OpenCL part.